src/ucp/proto/proto_debug.c

```
@@ -382,6 +430,207 @@ void ucp_proto_select_info_str(ucp_worker_h worker,
+        \
+        if (__perf_node == NULL) { \
+            return NULL; \
+        } \
+        \
+        __perf_node; \
+    })
+
+ucp_proto_perf_node_t *
+ucp_proto_perf_node_new_data(const char *name, const char *desc_fmt, ...)
+{
+    ucp_proto_perf_node_t *perf_node;
+
+    perf_node = UCP_PROTO_PERF_NODE_NEW(DATA, name, desc_fmt);
+    ucs_array_init_dynamic(&perf_node->data);
+    return perf_node;
+}
+
+ucp_proto_perf_node_t *ucp_proto_perf_node_new_select(const char *name,
+                                                      unsigned selected_child,
+                                                      const char *desc_fmt, ...)
+{
+    ucp_proto_perf_node_t *perf_node;
+
+    perf_node = UCP_PROTO_PERF_NODE_NEW(SELECT, name, desc_fmt);
+    perf_node->selected_child = selected_child;
+    return perf_node;
+}
+
+ucp_proto_perf_node_t *
+ucp_proto_perf_node_new_compose(const char *name, const char *desc_fmt, ...)
+{
+    return UCP_PROTO_PERF_NODE_NEW(COMPOSE, name, desc_fmt);
+}
+
+void ucp_proto_perf_node_ref(ucp_proto_perf_node_t *perf_node)
+{
+    if (perf_node != NULL) {
+        ++perf_node->refcount;
+    }
+}
+
+void ucp_proto_perf_node_deref(ucp_proto_perf_node_t **perf_node_p)
+{
+    ucp_proto_perf_node_t *perf_node = *perf_node_p;
+
+    if (perf_node == NULL) {
+        return;
+    }
+
+    ucs_assertv(perf_node->refcount > 0, "perf_node=%p name='%s' desc='%s'",
+                perf_node, perf_node->name, perf_node->desc);
+    --perf_node->refcount;
+    if (perf_node->refcount == 0) {
+        ucp_proto_perf_node_free(perf_node);
+        *perf_node_p = NULL;
+    }
+}
+
+static void
+ucp_proto_perf_node_append_child(ucp_proto_perf_node_t *perf_node,
+                                 ucp_proto_perf_node_t *child_perf_node)
+{
+    ucs_status_t status;
+
+    status = ucs_array_append(ucp_proto_perf_node, &perf_node->children);
+    if (status != UCS_OK) {
+        ucs_diag("failed to add perf node child");
+        return;
+    }
```

maybe replace by `ucs_assertv_always(status != UCS_OK, "failed to add perf node child: %s", ucs_status_string(status))`
normally it shouldn't happen

---

in perf_node we ignore errors related to memory allocation since the applications can work without having a performance information tree.